tools/wptrunner/wptrunner/executors/protocol.py

```diff
@@ -393,29 +393,32 @@ def switch_to_window(self, wptrunner_id, initial_window=None):
                 if not first or item != initial_window:
                     self.parent.base.set_window(item)
             else:
-                self._switch_to_frame(item)
-
-            try:
-                handle_window_id = self.parent.base.execute_script("return window.__wptrunner_id")
-                if str(handle_window_id) == wptrunner_id:
-                    return
-            except Exception:
-                pass
-            frame_count = self.parent.base.execute_script("return window.length")
-            if frame_count:
-                for frame_id in reversed(range(0, frame_count)):
-                    # None here makes us switch back to the parent after we've processed the frame
-                    stack.append(None)
-                    stack.append(frame_id)
+                try:
+                    self._switch_to_frame(item)
+                except ValueError:
+                    # The frame no longer exists, so continue
+                    pass
+
+            handle_window_id, nested_context_containers = self.parent.base.execute_script("""
+let contextParents = Array.from(document.querySelectorAll("frame, iframe, embed, object"))
+  .filter(elem => (elem.localName !== "embed" && elem.contentWindow) || elem.getSVGDocument());
```

Comment:
`HTMLFrameElement` [doesn't](https://html.spec.whatwg.org/multipage/obsolete.html#htmlframeelement) have a `getSVGDocument()` method.  I don't know how much we care about frames, but I wonder whether you meant `elem.localName !== "embed" ? elem.contentWindow : elem.getSVGDocument()`?

However, I wonder whether there is any way for `contentWindow` to be `null` on a container in a document with a non-null browsing context.
And [`getSVGDocument()`](https://html.spec.whatwg.org/multipage/embedded-content-other.html#dom-media-getsvgdocument) returns `null` if the SVG document is not same-origin.

The safe approach might be to be more generous here at the possible cost of more iterations in this loop.